staging/vhost-device-gpu/src/protocol.rs

```diff
@@ -0,0 +1,1438 @@
+    pub size: u32,
+    pub padding: u32,
+    pub edid: [u8; EDID_BLOB_MAX_SIZE],
+}
+
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for virtio_gpu_resp_edid {}
+
+// data passed in the control vq, 3d related
+
+#[derive(Copy, Clone, Debug, Default, FromBytes, AsBytes, PartialEq, Eq)]
+#[repr(C)]
+pub struct virtio_gpu_box {
+    pub x: u32,
+    pub y: u32,
+    pub z: u32,
+    pub w: u32,
+    pub h: u32,
+    pub d: u32,
+}
+
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for virtio_gpu_box {}
+
+// VIRTIO_GPU_CMD_TRANSFER_TO_HOST_3D, VIRTIO_GPU_CMD_TRANSFER_FROM_HOST_3D
+#[derive(Copy, Clone, Debug, Default, FromBytes, AsBytes, PartialEq, Eq)]
+#[repr(C)]
+pub struct virtio_gpu_transfer_host_3d {
+    pub box_: virtio_gpu_box,
+    pub offset: u64,
+    pub resource_id: u32,
+    pub level: u32,
+    pub stride: u32,
+    pub layer_stride: u32,
+}
+
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for virtio_gpu_transfer_host_3d {}
+
+// VIRTIO_GPU_CMD_RESOURCE_CREATE_3D
+pub const VIRTIO_GPU_RESOURCE_FLAG_Y_0_TOP: u32 = 1 << 0;
+#[derive(Copy, Clone, Debug, Default, FromBytes, AsBytes, PartialEq, Eq)]
+#[repr(C)]
+pub struct virtio_gpu_resource_create_3d {
+    pub resource_id: u32,
+    pub target: u32,
+    pub format: u32,
+    pub bind: u32,
+    pub width: u32,
+    pub height: u32,
+    pub depth: u32,
+    pub array_size: u32,
+    pub last_level: u32,
+    pub nr_samples: u32,
+    pub flags: u32,
+    pub padding: u32,
+}
+
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for virtio_gpu_resource_create_3d {}
+
+// VIRTIO_GPU_CMD_CTX_CREATE
+pub const VIRTIO_GPU_CONTEXT_INIT_CAPSET_ID_MASK: u32 = 1 << 0;
+#[derive(Copy, FromBytes, AsBytes, PartialEq, Eq)]
+#[repr(C)]
+pub struct virtio_gpu_ctx_create {
+    pub nlen: u32,
+    pub context_init: u32,
+    pub debug_name: [u8; 64],
+}
+
+// SAFETY: The layout of the structure is fixed and can be initialized by
+// reading its content from byte array.
+unsafe impl ByteValued for virtio_gpu_ctx_create {}
+
+impl Default for virtio_gpu_ctx_create {
+    fn default() -> Self {
+        // SAFETY: trivially safe
+        unsafe { ::std::mem::zeroed() }
+    }
+}
+
+impl Clone for virtio_gpu_ctx_create {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl fmt::Debug for virtio_gpu_ctx_create {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let debug_name = CStr::from_bytes_with_nul(&self.debug_name[..min(64, self.nlen as usize)])
+            .map_or_else(
+                |err| format!("Err({})", err),
+                |c_str| c_str.to_string_lossy().into_owned(),
+            );
```

Comment:
I would make the field private & create a getter for this?

Reply:
Getter sounds good, but I would still keep the field `pub`. 
This also implements `FromBytes` so I don't think hiding fields accomplishes anything + the field being public is also useful for tests.